### PR TITLE
Allow remotes to not open a legacy substream

### PR DESCRIPTION
**Pull request by tomaka**
This tackles the first bullet point of this comment: https://github.com/paritytech/substrate/issues/5670#issuecomment-690117333

At the moment, when a connection is established and peerset slots reserved, nodes have to open a legacy substream and all notification substreams. Once all these substreams are open, the actual opening is reported on the API layer.

Amongst the notification substreams, the block announces substream contains a handshake that mimics the handshake of the legacy substream.

This PR changes the code so that we report the actual opening if all notification substreams are open, even if the legacy substream is closed. In that situation, the handshake of the block announces substream is reported as the handshake.

Since the handshake actually differ a bit (there is no protocol version system anymore), I've done two changes:

- In `protocol.rs`, we now try to parse the handshake as either the legacy handshake or the block announces handshake.
- The `system_peers` JSON-RPC call no longer has a `protocol_version` field. I believe that this shouldn't have been exposed in the first place.

I'm going to open another pull request on top of this one that changes the code to no longer open a legacy substream altogether. EDIT: https://github.com/paritytech/substrate/pull/7076 We should merge this PR only if the CI of that other pull request is green. However, we need to publish a version between this PR and that other PR.
Nodes having both PRs will no longer be able to connect to nodes that have neither. The opposite stills works: nodes without these two PRs will still be able to connect to nodes with both. As such, this change is considered acceptable.

### CODE REVIEW

File: client/network/src/protocol.rs [hunk 1]
```diff
@@ -419,12 +408,23 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
...
+			let block_announces_handshake = BlockAnnouncesHandshake::build(&config, &chain).encode();
+			GenericProto::new(
+				local_peer_id,
+				protocol_id.clone(),
+				versions,
+				build_status_message(&config, &chain),
+				peerset,
+				iter::once((block_announces_protocol.clone(), block_announces_handshake))
+					.chain(iter::once((transactions_protocol.clone(), vec![]))),
+				0,
```
[the remark this answers is not in the document]
Comment (tomaka) [reply client/network/src/protocol.rs:424]: > have the Iterator produce tuples of both protocol name as well as a bool indicating whether it is the handshake protocol?

Then we need to enforce that one and only one element in the iterator contains `true`.

[the remark this answers is not in the document]
Comment (mxinden) [reply client/network/src/protocol.rs:424]: I am not saying it is a good solution, only that I find it less error prone than passing an index.

Is there ever the use case to not have the first protocol be the one providing the handshake? If not, I would just always interpret the very first passed notifications protocol as the one providing the handshake. What do you think?

[the remark this answers is not in the document]
Comment (tomaka) [reply client/network/src/protocol.rs:424]: > Is there ever the use case to not have the first protocol be the one providing the handshake? If not, I would just always interpret the very first passed notifications protocol as the one providing the handshake. What do you think?

It's not really a matter of use-case. The caller chooses the order of the protocols, and can always move the handshake-bearing protocol as the first one.
However I would find it more confusing, as the fact that we special-case a certain protocol would no longer be explicited through the parameters.

[the remark this answers is not in the document]
Comment (mxinden) [reply client/network/src/protocol.rs:424]: I don't have a strong opinion on this. To me tracking both a list of protocols as well as an index into that list is error prone. I think that tracking a `bool` alongside each protocol is less error prone.

**Review by mxinden — APPROVED** [review]

**Comment by tomaka** [comment]
Nothing weird happening on the burn-in. The node is surprisingly better connected than it was before its reboot, but version that was on the node was quite old and we can't attribute this change to this pull request.

**Comment by mxinden** [comment]
> The node is surprisingly better connected than it was before its reboot, but the version that was on the node was quite old and we can't attribute this change to this pull request.

In case you want ot compare to an up-to-date `master` deployment, `canary-sentry-ew4-0` would be an option.

File: client/network/src/protocol.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomaka) [reply client/network/src/protocol.rs:424]: As mentioned I'm of the opposite opinion :sweat_smile: 
While it makes the API nicer-looking, you then introduce the possible situation where two protocols accidentally have the flag set to true.

Also, I know it's not great to say this, but this code is temporary and will be removed again once substreams can be opened individually.

[the remark this answers is not in the document]
Comment (romanb) [reply client/network/src/protocol.rs:424]: > Is there ever the use case to not have the first protocol be the one providing the handshake? If not, I would just always interpret the very first passed notifications protocol as the one providing the handshake. 

I'd prefer this as well, especially if the special-casing of the first protocol is temporary.

**Review by romanb — APPROVED** [review]

**Comment by tomaka** [comment]
I did the change to have the first protocol always containing the handshake.

**Comment by tomaka** [comment]
bot merge

**Pull request merged**